clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java

```diff
@@ -69,15 +70,22 @@ RecordMetadata valueOrError() throws ExecutionException {
     
     RecordMetadata value() {
         return new RecordMetadata(result.topicPartition(), this.result.baseOffset(), this.relativeOffset,
-                                  this.timestamp, this.checksum, this.serializedKeySize, this.serializedValueSize);
+                                  timestamp(), this.checksum, this.serializedKeySize, this.serializedValueSize);
     }
     
     public long relativeOffset() {
         return this.relativeOffset;
     }
 
+    public long createTimestamp() {
+        return this.createTimestamp;
+    }
+
+    /**
+     * If the timestamp returned by server is NO_TIMESTAMP, CreateTime is being used. Otherwise LogAppendTime is being used.
+     */
     public long timestamp() {
-        return this.timestamp;
+        return result.responseTimestamp() == Record.NO_TIMESTAMP ? createTimestamp() : result.responseTimestamp();
```

Comment:
It seems difficult to do that since `PartitionResponse` doesn't have the create timestamp. I tried to make things clearer in my last commit. Please take a look.